File: backend/schemas/fermentables_schema.py
from marshmallow import Schema, fields, validates, ValidationError

class FermentablesSchema(Schema):
    itemUserId = fields.Int(load_only=True)
    id = fields.Int(dump_only=True)
    name = fields.Str(required=True)
    description = fields.Str(allow_none=True)
    ebc = fields.Decimal(required=True, as_string=True)
    potentialExtract = fields.Decimal(required=True, as_string=True)
    type = fields.Str(required=True)
    supplier = fields.Str(required=True)
    unitPrice = fields.Decimal(allow_none=True, as_string=True)

    @validates("name")
    def validate_name(self, value, *args, **kwargs):
        if not value.strip():
            raise ValidationError("Name must not be empty.")
        if len(value) > 40:
            raise ValidationError("Name must be at most 40 characters long.")

    @validates("ebc")
    def validate_ebc(self, value, *args, **kwargs):
        if value < 0:
            raise ValidationError("EBC must be a positive number.")

    @validates("potentialExtract")
    def validate_potential_extract(self, value, *args, **kwargs):
        if value < 0 or value > 2.000:
            raise ValidationError("Potential extract must be between 0 and 2.000.")

    @validates("type")
    def validate_type(self, value, *args, **kwargs):
        if value.strip() not in ['Base', 'Specialty', 'Adjunct']:
            raise ValidationError("Invalid type. Must be one of: Base, Specialty or Adjunct.")

    @validates("supplier")
    def validate_supplier(self, value, *args, **kwargs):
        if len(value) > 40:
            raise ValidationError("Supplier name must be at most 40 characters long.")

    @validates("unitPrice")
    def validate_unit_price(self, value, *args, **kwargs):
        if value is not None and value < 0:
            raise ValidationError("Unit price must be greater than or equal to 0.")

File: frontend/src/Pages/Recipes/schemas/fermentablesSchema.js
import * as yup from 'yup';

export const addFermentableSchema = yup.object().shape({
  selectedItem: yup.object().required('Please select an item'),
  quantity: yup
    .number()
    .typeError('Quantity must be a number')
    .positive('Quantity must be greater than zero')
    .required('Quantity is required'),
});

export const updateFermentableSchema = yup.object().shape({
  name: yup
    .string()
    .trim()
    .required('Name is required')
    .min(2, 'Name must be at least 2 characters'),
  description: yup
    .string()
    .nullable(),
  type: yup
    .string()
    .oneOf(['base', 'especial', 'adjunct'], 'Invalid fermentable type')
    .required('Fermentable Type is required'),
  supplier: yup
    .string()
    .nullable(),
  ebc: yup
    .number()
    .typeError('EBC must be a number')
    .min(0, 'EBC cannot be negative')
    .nullable(),
  potentialExtract: yup
    .number()
    .typeError('Potential Extract must be a number')
    .min(0, 'Potential Extract cannot be negative')
    .nullable(),
  quantity: yup
    .number()
    .typeError('Quantity must be a number')
    .positive('Quantity must be greater than zero')
    .required('Quantity is required'),
});
